File: server/internal/catalog/keyparse.go
package catalog

import (
	"regexp"
	"strings"
)

// keyparse.go is the Go mirror of the Python builder's
// keyparse.parse_hive_key (mcap_catalog/mcap_catalog_builder/keyparse.py).
// Its invariant is rebuildHiveKey(parseHiveKey(k)) == strings.TrimLeft(k, "/")
// — an inverse of rebuildHiveKey up to stripped leading slashes, matching the
// Python side's key.lstrip("/") (see parseHiveKey's doc comment below). It
// exists for the tag-IPC forwarder (D2, tagforward.go): the tag-edit IPC
// endpoint addresses files by object key, not file_id (CATALOG_CONTRACT.md
// §10 "Key-based addressing"), so the forwarder must turn a wire file_id into
// a key (rebuildHiveKey, already landed) and — after the edit — a key back
// into dimension NAMES to re-resolve the CURRENT file_id (this file).
var hiveKeyRe = regexp.MustCompile(
	`^customer=([^/]+)/customer_site=([^/]+)/robot=([^/]+)/source=([^/]+)/date=([^/]+)/([^/]+\.mcap)$`)

// hiveDims is one Hive key's parsed dimension NAMES (not catalog row ids —
// the caller resolves those itself, e.g. by joining on these names).
type hiveDims struct {
	Customer, Site, Robot, Source, Date, Filename string
}

// parseHiveKey parses key into its dimension names. The invariant is
// rebuildHiveKey(parseHiveKey(k)) == strings.TrimLeft(k, "/") for every k this
// accepts — NOT a plain "== k": key's leading slashes (if any) are stripped
// first, mirroring the Python side's `key.lstrip("/")` (ALL leading slashes,
// not just one), so a key that had leading slashes round-trips to its
// slash-stripped form, not to itself verbatim. ok is false if key does not
// match the template exactly (a flat name, a missing partition, a
// non-".mcap" filename, an extra path segment, …).
func parseHiveKey(key string) (hiveDims, bool) {
	m := hiveKeyRe.FindStringSubmatch(strings.TrimLeft(key, "/"))
	if m == nil {
		return hiveDims{}, false
	}
	return hiveDims{
		Customer: m[1],
		Site:     m[2],
		Robot:    m[3],
		Source:   m[4],
		Date:     m[5],
		Filename: m[6],
	}, true
}

File: mcap_catalog/mcap_catalog_builder/keyparse.py
"""Hive object-key parsing, rebuilding, and relative-path key derivation.

The object key is a Hive-partitioned path::

    customer=<c>/customer_site=<site>/robot=<r>/source=<s>/date=<d>/<filename>.mcap

Note the path literal is ``customer_site=`` but the parsed field is ``site``.
``rebuild_hive_key`` is the exact inverse of ``parse_hive_key`` so the catalog builder
can verify ``rebuild_hive_key(dims) == key.lstrip('/')`` before trusting a parse.
"""

import os
import re

HIVE_RE: re.Pattern = re.compile(
    r"^customer=(?P<customer>[^/]+)/"
    r"customer_site=(?P<site>[^/]+)/"
    r"robot=(?P<robot>[^/]+)/"
    r"source=(?P<source>[^/]+)/"
    r"date=(?P<date>[^/]+)/"
    r"(?P<filename>[^/]+\.mcap)$"
)


def parse_hive_key(key: str) -> dict[str, str] | None:
    """Parse a Hive key into ``{customer, site, robot, source, date, filename}``.

    Returns ``None`` if the key does not match the exact template (e.g. a flat
    name, a missing partition, or a non-``.mcap`` filename).
    """
    m = HIVE_RE.match(key.lstrip("/"))
    return m.groupdict() if m is not None else None


def rebuild_hive_key(dims: dict[str, str]) -> str:
    """Rebuild the object key from parsed dimensions (exact inverse of parse)."""
    return (
        f"customer={dims['customer']}/"
        f"customer_site={dims['site']}/"
        f"robot={dims['robot']}/"
        f"source={dims['source']}/"
        f"date={dims['date']}/"
        f"{dims['filename']}"
    )


def relpath_key(abs_path: str, watched_root: str) -> str:
    """The file's path relative to ``watched_root``, POSIX-normalized, no leading slash."""
    return os.path.relpath(abs_path, watched_root).replace(os.sep, "/")
